spec.bs

```diff
@@ -13,3 +13,540 @@ Group: patcg
+  Topics can be send in the HTTP header for {{WindowOrWorkerGlobalScope/fetch()}} request. The response header for a topics related request can specify whether the caller should to be recorded.
+
+  <h3 id="send-browsing-topics-header-boolean-associated-with-request-header">send browsing topics header boolean associated with Request</h3>
+  A [=request=] has an associated <dfn for=request>send browsing topics header boolean</dfn>. Unless stated otherwise it is unset.
+
+  <span class=XXX>TODO: make the modification directly to the fetch spec.</span>
+
+  <h3 id="browsing-topics-attribute-in-request-init-header">browsingTopics attribute in RequestInit</h3>
+  The <a href="https://fetch.spec.whatwg.org/#requestinit">RequestInit</a> dictionary contains a browsingTopics attribute:
+
+  <pre class=idl>
+  dictionary RequestInit {
+    // existing attributes ...
+
+    boolean browsingTopics;
+  };
+  </pre>
+
+  <span class=XXX>TODO: make the modification directly to the fetch spec.</span>
+
+  <h3 id="modification-to-request-constructor-steps-header">Modification to request constructor steps</h3>
+  The following step will be added to the <a href="https://fetch.spec.whatwg.org/#dom-request">new Request(input, init) constructor steps</a>, before step "Set this's [=Request/request=] to |request|":
+
+  1. If <var ignore=''>init</var>["{{RequestInit/browsingTopics}}"] <a for=map>exists</a>, then set |request|'s [=request/send browsing topics header boolean=] to it.
+
+  <span class=XXX>TODO: make the modification directly to the fetch spec.</span>
+
+  <h3 id="the-sec-browsing-topics-http-request-header-header">The \`<code>Sec-Browsing-Topics</code>\` HTTP request header</h3>
+
+  This specification defines a \`<dfn export http-header><code>Sec-Browsing-Topics</code></dfn>\` HTTP request header. It is used to send the topics.
+
+  <h3 id="modification-to-http-network-or-cache-fetch-algorithm-header">Modification to HTTP-network-or-cache fetch algorithm</h3>
+  The following step will be added to the <a href="https://fetch.spec.whatwg.org/#concept-http-network-or-cache-fetch">HTTP-network-or-cache fetch</a> algorithm, before step "Modify |httpRequest|'s [=request/header list=] per HTTP. ...":
+
+  1. <a>Append or modify a request \`<code>Sec-Browsing-Topics</code>\` header</a> for |httpRequest|.
+
+  <span class=XXX>TODO: make the modification directly to the fetch spec.</span>
+
+  <h3 id="append-or-modify-a-request-sec-browsing-topics-header-header">Append or modify a request `Sec-Browsing-Topics` header</h3>
+
+  <div algorithm>
+    To <dfn>append or modify a request \`<code>Sec-Browsing-Topics</code>\` header</dfn>, given a [=request=] |request|, run these steps:
+      1. If |request|'s [=request/send browsing topics header boolean=] is not true, then return.
+      1. [=header list/Delete=] [:Sec-Browsing-Topics:] from |request|’s [=header list=].
```

Comment:
@domenic (for some reason I'm unable to comment on the last thread) I think the origin itself can just change? (i.e. cross-origin redirect, from exampleA.com to exampleB.com?) This should be the main reason why we would send different headers.

The timestamp does matter as well but should be a minor thing. For a single request-redirect chain, this is very unlikely to cause the result to be different, though possible. So I'm not sure if we need to mention this (esp. if the previous point is valid and makes it obvious enough that the headers can be different).

Reply:
Yes, that's also possible. But I wanted to confirm you also wanted to change it in the same-origin case. If you only wanted to change it cross-origin, then adding a guard like https://fetch.spec.whatwg.org/#ref-for-cors-non-wildcard-request-header-name does would be more what I expect.

I still think it's uncommon enough to delete headers, that you should write a note explaining why you're doing so. That note can mention the cross-origin case as an especially clear example of why, but the reader might then ask the same question I'm asking, which is "why not only remove in the cross-origin case?" So I think explaining that even for same-origin redirects you want to recalculate the header, would be helpful.

Reply:
Ack. Done.